src/qt/treectrl.cpp

```diff
@@ -125,6 +125,61 @@ class ImageState
 
 Q_DECLARE_METATYPE(TreeItemDataQt)
 
+// We use a custom selection model because we need to generate the
+// wxEVT_TREE_SEL_CHANGED event only if it is allowed. that is, only
+// if the wxEVT_TREE_SEL_CHANGING handler returns true.
+class wxQItemSelectionModel : public QItemSelectionModel
+{
+public:
+    wxQItemSelectionModel(QAbstractItemModel* model, QObject* parent)
+        : QItemSelectionModel(model, parent)
+    {}
+
+    virtual void
+    select(const QItemSelection &selection, QItemSelectionModel::SelectionFlags command)
+    {
+        if ( ms_isSelectionChangeAllowed )
+        {
+            QItemSelectionModel::select(selection, command);
+        }
+    }
+
+    virtual void
+    select(const QModelIndex &index, QItemSelectionModel::SelectionFlags command) override
+    {
+        if ( ms_isSelectionChangeAllowed )
+        {
+            QItemSelectionModel::select(index, command);
+        }
+    }
+
+    virtual void
+    setCurrentIndex(const QModelIndex& index, QItemSelectionModel::SelectionFlags command) override
+    {
+        ms_isSelectionChangeAllowed = IsSelectionChangeAllowed(index);
+
+        if ( ms_isSelectionChangeAllowed )
+        {
+            QItemSelectionModel::setCurrentIndex(index, command);
+        }
+    }
+
+private:
+    bool IsSelectionChangeAllowed(const QModelIndex& index) const;
+
+    static bool ms_isSelectionChangeAllowed;
+};
+
+/*static*/
+bool wxQItemSelectionModel::ms_isSelectionChangeAllowed = true;
+
+namespace
+{
+// Used when generating wxEVT_TREE_SEL_{CHANGING,CHANGED} events only
+static wxTreeItemId gs_currentItem;
+static wxTreeItemId gs_previousItem;
```

Comment:
Do these really need to be global? I am not 100% sure it's not going to be a problem, e.g. if a handler of some event changes selection in some other tree control.